kernel/arch/dreamcast/hardware/sq.c

```diff
@@ -115,38 +122,48 @@ void sq_wait(void) {
 
 /* Copies n bytes from src to dest, dest must be 32-byte aligned */
 __attribute__((noinline)) void *sq_cpy(void *dest, const void *src, size_t n) {
-    uint32_t *d = SQ_MASK_DEST(dest);
     const uint32_t *s = src;
+    void *curr_dest = dest;
+    uint32_t *d;
+    size_t nb;
 
     /* Fill/write queues as many times necessary */
     n >>= 5;
 
-    /* Exit early if we dont have enough data to copy */
-    if(n == 0)
-        return dest;
-
-    sq_lock(dest);
-
-    /* If src is not 8-byte aligned, slow path */
-    if ((uintptr_t)src & 7) {
-        while(n--) {
-            dcache_pref_block(s + 8); /* Prefetch 32 bytes for next loop */
-            d[0] = *(s++);
-            d[1] = *(s++);
-            d[2] = *(s++);
-            d[3] = *(s++);
-            d[4] = *(s++);
-            d[5] = *(s++);
-            d[6] = *(s++);
-            d[7] = *(s++);
-            sq_flush(d);
-            d += 8;
+    while (n > 0) {
+        /* Transfer maximum 512 KiB at once. This is because when using the
+         * MMU the SQ area is a 1 MiB page, and the destination address may
+         * not be on a page boundary. */
+        nb = n > 0x4000 ? 0x4000 : n;
+
+        d = sq_lock(curr_dest);
+
```

Comment:
With the changes I suggested for adding a new function to gather this, it shouldn't be necessary to lock/unlock at every 512 loop.

Reply:
Yeah, this change may slow down copying large textures.
But I like the return of the destination, it should have been done right away.

Reply:
> Yeah, this change may slow down copying large textures. But I like the return of the destination, it should have been done right away.
> 
> Edit: Although if you look at how I rewrote [convert_YUV420_to_YUV422_texture](https://github.com/KallistiOS/KallistiOS/blob/master/examples/dreamcast/pvr/yuv_converter/YUV420/yuv420.c#L183) in the example, you can understand that the single destination is a rarity :)) snd_pcm16_split_sq is lite case.

Yea, I was originally going to suggest to not return the address and keep it as void adding the macro replacement to get the address. But I see too how it simplifies the most common uses and doesn't interfere with these more complicated ones :)

Reply:
> With the changes I suggested for adding a new function to gather this, it shouldn't be necessary to lock/unlock at every 512 loop.

It would still be needed, as you might have to reprogram the MMU SQ window.